pkg/activator/requestcontrol/activator.go

```diff
@@ -0,0 +1,295 @@
+	logger := log.FromContext(ctx)
+	namespace := pool.Namespace
+
+	// TODO: store annotation values in datastore to avoid reading them every time
+	// verify required inferencePool annotations
+	valid := verifyPoolObjectAnnotations(logger, pool)
+	if !valid {
+		return false
+	}
+
+	// extract optional inferencePool annotation if it exists, otherwise use a default value
+	scaleGracePeriod := DefaultScaleFromZeroGracePeriod
+	if value, found := getOptionalPoolAnnotation(logger, scaleFromZeroGracePeriodKey, pool); found {
+		scaleGracePeriod, _ = time.ParseDuration(value)
+	}
+
+	// Get the scale subresource for the target inferencePool object
+	gvr, err := GetResourceForKind(a.Mapper, pool.Annotations[objectAPIVersionKey], pool.Annotations[objectkindKey])
+	if err != nil {
+		msg := "Failed to parse Group, Version, Kind, Resource"
+		logger.Error(err, msg, "apiVersion", pool.Annotations[objectAPIVersionKey], "kind", pool.Annotations[objectkindKey])
+		return false
+	}
+
+	gr := gvr.GroupResource()
+	scaleObject, err := a.ScaleClient.Scales(namespace).Get(ctx, gr, pool.Annotations[objectNameKey], metav1.GetOptions{})
+	if err != nil {
+		logger.Error(err, "Error getting scale subresource object")
+		return true
+	}
+
+	// Common case: enough replicas?
+	if scaleObject.Spec.Replicas > 0 {
+		if a.inferencePoolPodsReady(logger, namespace, pool.Annotations[objectNameKey], scaleObject.Spec.Replicas, scaleGracePeriod, gvr) {
+			// Scale object exists and has no zero running replicas then do not scale it
+			logger.V(logutil.DEBUG).Info(fmt.Sprintf("Scale Object %s have at least one replica ready. Skipping scaling from zero", scaleObject.Name))
+			return true
+		}
+	}
+
+	// Need to scale inferencePool workload from zero to one replicas
+	numReplicas := int32(1)
+	scaleData := scaledObjectData{name: pool.Annotations[objectNameKey], scaleGracePeriod: DefaultScaleFromZeroGracePeriod, numReplicas: numReplicas, scaleObject: scaleObject}
+	return a.scaleInferencePool(ctx, logger, namespace, scaleData, gr, gvr)
+}
+
+func (a *Activator) inferencePoolPodsReady(logger logr.Logger, namespace, objname string, numReplicas int32, scaleGracePeriod time.Duration, gvr schema.GroupVersionResource) bool {
+	// Don't inherit the parent context to avoid cancellation
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, scaleGracePeriod, false, func(ctx context.Context) (done bool, err error) {
+
+		a.datastore.ResetTicker(DefaultScaleDownDelay) // turn off the deactivator during scale from zero events
+
+		unstructuredObj, err := a.DynamicClient.Resource(gvr).Namespace(namespace).Get(ctx, objname, metav1.GetOptions{})
+		if err != nil {
+			logger.Error(err, "Error getting unstructured object")
+			return false, nil // continue polling
+		}
+
+		// NOTE: this assumes that the target object has a status.readyReplicas field
+		if readyReplicas, ok := unstructuredObj.Object["status"].(map[string]any)["readyReplicas"].(int64); ok {
```

Comment:
should we use `readyReplicas, ok, err := unstructured.NestedFieldNoCopy(unstructuredObj, "status", "readyReplicas")`?